Indicators/TimeSeriesIndicator.cs

```diff
@@ -0,0 +1,158 @@
+/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace QuantConnect.Indicators
+{
+    /// <summary>
+    /// The base class for any Time Series-type indicator, containing methods common to most of such models.
+    /// </summary>
+    public abstract class TimeSeriesIndicator : IndicatorBase<IndicatorDataPoint>, IIndicatorWarmUpPeriodProvider
+    {
+        private double[] _diffHeads;
+
+        /// <summary>
+        /// A constructor for a basic Time Series indicator.
+        /// </summary>
+        /// <param name="name">The name of this indicator</param>
+        protected TimeSeriesIndicator(string name)
+            : base(name)
+        {
+        }
+
+        /// <summary>
+        /// Gets a flag indicating when this indicator is ready and fully initialized
+        /// </summary>
+        public override bool IsReady { get; }
+
+        /// <summary>
+        /// Required period, in data points, for the indicator to be ready and fully initialized.
+        /// </summary>
+        public int WarmUpPeriod { get; }
+
+        /// <summary>
+        /// Computes the next value of this indicator from the given state
+        /// </summary>
+        /// <param name="input">The input given to the indicator</param>
+        /// <returns>A new value for this indicator</returns>
+        protected override decimal ComputeNextValue(IndicatorDataPoint input)
+        {
+            return 0m;
+        }
+
+        /// <summary>
+        /// Differences a time series d times.
+        /// </summary>
+        /// <param name="series">Series to difference</param>
+        /// <param name="d">The differencing order</param>
+        protected double[]
+            DifferenceSeries(int d, double[] series)
+        {
+            _diffHeads = new double[series.Length - d];
+            if (d == 0)
+            {
+                return null;
+            }
+
+            var localSeries = series;
+            for (var j = 1; j <= d; j++)
```

Comment:
Wondering if some of these methods couldn't be replaced by a call to `Math.` or `MathNet.Numerics.` for example, maybe it's already implemented?

Let's please add a couple of specific unit tests for each of these different methods 👍. Feels like the core implementation could be a static methods? Easier to test and reuse.

`var j = 1` starting from `0` would avoid the `-1` bellow?
nit - let's remove the new line after `protected double[]`, same bellow at `InverseDifferencedSeries`

Reply:
Unfortunately the implemented Math.Net does not offer much in terms of Time Series. Newer versions have implemented some useful functions however (e.g. https://github.com/mathnet/mathnet-numerics/blob/64aa01e33c10b5063057128739d60c34ddc373ea/src/Numerics/Statistics/Correlation.cs#L46)

Good idea w.r.t. implementing tests for these -- probably should have had a TimeSeriesIndicator test file!  

Lastly, the `var j  = 1` can be changed such that `j + 1 <= d`.

Reply:
hm this static method isn't being used and it's duplicated bellow?

Reply:
I gave it a static method and a non-static implementation such that one can utilize it outside of the context of indicators (and thus, for example, allowing us to use it in a test where we aren't concerned with initializing an ARIMA).
e.g., 
```
  double[] heads;
            var differencer = TimeSeriesIndicator.DifferenceSeries(1, test, out heads);        
```